pkg/render/common/components/components.go

```diff
@@ -0,0 +1,176 @@
+	var nodeSelector map[string]string
+	var initContainers []container
+	var containers []container
+
+	switch obj := overrides.(type) {
+	case *operatorv1.CalicoNodeDaemonSet:
+		// If overrides was a pointer to nil, then just return the original ds.
+		if obj == nil {
+			return ds
+		}
+
+		metadata = obj.Metadata
+
+		// Skip the rest of the case if no spec.
+		if obj.Spec == nil {
+			break
+		}
+
+		minReadySeconds = obj.Spec.MinReadySeconds
+
+		// Skip the rest of the case if no pod template spec.
+		if obj.Spec.Template == nil {
+			break
+		}
+
+		podTemplateMetadata = obj.Spec.Template.Metadata
+
+		if obj.Spec.Template.Spec != nil {
+			affinity = obj.Spec.Template.Spec.Affinity
+			tolerations = obj.Spec.Template.Spec.Tolerations
+			nodeSelector = obj.Spec.Template.Spec.NodeSelector
+
+			if len(obj.Spec.Template.Spec.InitContainers) > 0 {
+				for _, v := range obj.Spec.Template.Spec.InitContainers {
+					c := container{Name: v.Name, Resources: v.Resources}
+					initContainers = append(initContainers, c)
+				}
+			}
+			if len(obj.Spec.Template.Spec.Containers) > 0 {
+				for _, v := range obj.Spec.Template.Spec.Containers {
+					c := container{Name: v.Name, Resources: v.Resources}
+					containers = append(containers, c)
+				}
+			}
+		}
+	}
+
+	if metadata != nil {
+		if len(metadata.Labels) > 0 {
+			if ds.Labels == nil {
+				ds.SetLabels(make(map[string]string))
+			}
+			mergeMaps(ds.Labels, metadata.Labels)
+		}
+		if len(metadata.Annotations) > 0 {
+			if ds.GetAnnotations() == nil {
+				ds.SetAnnotations(make(map[string]string))
+			}
+			mergeMaps(ds.GetAnnotations(), metadata.Annotations)
+		}
+	}
+
+	if minReadySeconds != nil {
+		ds.Spec.MinReadySeconds = *minReadySeconds
+	}
+
+	if podTemplateMetadata != nil {
+		if len(podTemplateMetadata.Labels) > 0 {
+			if ds.Spec.Template.GetLabels() == nil {
+				ds.Spec.Template.SetLabels(make(map[string]string))
+			}
+			mergeMaps(ds.Spec.Template.GetLabels(), podTemplateMetadata.Labels)
+		}
+		if len(podTemplateMetadata.Annotations) > 0 {
+			if ds.Spec.Template.GetAnnotations() == nil {
+				ds.Spec.Template.SetAnnotations(make(map[string]string))
+			}
+			mergeMaps(ds.Spec.Template.GetAnnotations(), podTemplateMetadata.Annotations)
+		}
+	}
+
+	// Merge all other fields in the overrides spec if they exist.
+	if initContainers != nil {
+		mergeContainers(ds.Spec.Template.Spec.InitContainers, initContainers)
+	}
+	if containers != nil {
+		mergeContainers(ds.Spec.Template.Spec.Containers, containers)
+	}
+	if affinity != nil {
+		ds.Spec.Template.Spec.Affinity = affinity
+	}
+	if nodeSelector != nil {
+		ds.Spec.Template.Spec.NodeSelector = nodeSelector
+	}
+	if tolerations != nil {
+		ds.Spec.Template.Spec.Tolerations = tolerations
+	}
+
+	return ds
+}
+
+// mergeMaps adds all key/value pairs from the provided map to the current map
+// if those keys do not already exist in the current map.
+func mergeMaps(current map[string]string, provided map[string]string) {
```

Comment:
I believe we have a function in another package that does this already - maybe should make that one public.

Reply:
https://github.com/tigera/operator/blob/7de6f2820556f766aed93eded7ef1f2895f980f7/pkg/controller/utils/component.go#L527-L537